### ISSUE TITLE
projectile-find-tag missing tags

### ISSUE THREAD

**Issue by fommil** [bbatsov/projectile#683]
Comparing against http://emacswiki.org/emacs/InteractivelyDoThings#toc11

```elisp
  (defun my-ido-find-tag ()
    "Find a tag using ido"
    (interactive)
    (tags-completion-table)
    (let (tag-names)
      (mapatoms (lambda (x)
                  (push (prin1-to-string x t) tag-names))
                tags-completion-table)
      (find-tag (ido-completing-read "Tag: " tag-names))))
```

the `projectile-find-tag` is missing a lot of entries on a Java/Scala project. e.g. classnames don't appear to be searched with projectile.

**Comment by rootzlevel** [comment]
I can confirm this issue.

To reproduce, try to jump to the tag `Dispatcher` with projectile and [this TAGS file](http://wwwcip.cs.fau.de/~lu03pevi/TAGS). It is not offered as one of the alternatives.

`M-x find-tag` on the same tag works just fine.

My emacs version is 24.5.1 and projectile is of version 20150428.158

**Comment by fommil** [comment]
I was looking into this again and found out that the `ido` completions do not find the tag, but if I type out what I'm searching for in full, then it does seem to be able to jump to the definition. Perhaps `ido` is just aggressively pruning and dropping valid completions?

**Comment by rootzlevel** [comment]
It's not `ido`, setting `projectile-completion-system` to `'default` has the same problem.

**Comment by fommil** [comment]
ok, well I'm just using `find-tag` as the binding until this is fixed so that's a reasonable workaround.

**Comment by base698** [comment]
I'm having this problem too with a javascript project.

**Comment by jaycotton** [comment]
I'm having the same problem in both Perl and Python projects. `projectile-find-tag` does not include many function names in the completion list. Looking into it a bit more, I see that the completion list is being generated by the function etags-tags-completion-table (on my setup at least), from the etags emacs package.

I can manually complete the function I'm looking for, but if there are multiple instances of that function definition in the project, it just jumps to the first one rather than requesting disambiguation.

**Comment by jaycotton** [comment]
The problem is that the completion table returned by `etags-tags-completion-table` has a hard-coded length of 511. Only the last 511 tags in the TAGS file will be returned. My file has around 11,000 tags.

It is surprising to me that `find-tag`, which is in the same package as `etags-tags-completion-table`, does not share this limitation. It happily provides me with 11,000 items in the completion buffer.

**Comment by base698** [comment]
Good find.  If you change the limit does it work?

**Comment by jaycotton** [comment]
Yes, it works, but I'm not satisfied with a hard coded solution. I feel like I'm missing something. How is this not a major problem for almost all users?

I will poke around a bit more when I have time. I'll post it here if I find a solution.

**Comment by jaycotton** [comment]
To be clear, this limitation is inside the etags package, not projectile, although projectile depends on etags.

**Comment by jaycotton** [comment]
I was wrong when I described the hard-coded length of the completion table as a *limitation*. The table is actually an `obarray`. Its length is a number of `buckets`, and each bucket may contain many hashed symbols. The way this `obarray` was initialized with length 511 (one less than a power of two) is actually the recommended approach.

So, the table does in fact have all of the symbols, but some downstream logic iterates over the buckets, probably treating the `obarray` like a normal sequence, so we end up with only 511 completion entries.

The only way to get all the symbols out of an `obarray` (as far as I know) is by using the `mapatoms` function. I tried altering the `projectile--tags` function to do this, but it was very, very slow.

As mentioned above, the `find-tags` function in the etags package, called interactively with no arguments, does have completion with the entire collection of tags. It does this by passing the completions table to `all-completions`, which is a primitive C function, possibly for the efficiency reasons I encountered when I tried to call `mapatoms` myself in elisp.

The following snippet of advice can be added to your .emacs and will provide the functionality we want (at a small performance penalty on my machine). I'm sure there is a better way. I'll post it if I come up with it.

```
(defun my-expand-completion-table (orig-fun &rest args)
  "Extract all symbols from COMPLETION-TABLE before calling projectile--tags."
  (let ((completion-table (all-completions "" (car args))))
    (funcall orig-fun completion-table)))
  
(advice-add 'projectile--tags :around #'my-expand-completion-table)
```
